### PR TITLE
Fudan FM11R08* full-card recovery script; with Bambu tag support

### PR DESCRIPTION
**Pull request by csBlueChip**
Beyond the original recovery script (used by this script) this script performs a number of tasks to aid us teaching NFC at our local defcon meet.
qv. https://eprint.iacr.org/2024/1275.pdf [Philippe Teuwen]

+ Full breakdown of Block 0 (esp. Fudan specific details)
+ Verify UID/BCC
+ SAK decode
+ Optionally perform online Fudan validation check
+ [[key recovery]]
+ Dump all 64+8 blocks
+ ...with all 32+2 keys
+ Decode ACL [clearer output]
+ [[MAD decode]]
+ Bambu tag decode
+ Create full 64+8 block dump (incl keys) ...for the CTF flag

### CODE REVIEW

**Comment by iceman1001** [comment]
Nice one,
We prefer to use GPL3.0 or later as a license in this project,   so if you wouldn't mind?

**Comment by csBlueChip** [comment]
I totally failed to spot the GPL3 licence - clearly monoscopically focused
on getting the script working - <facepalm>

If I change my script to be dual licence (GPL/MIT), will that satisfy your
requirements? ...Or is it a case of "GPL3 or nothing"?

File: client/pyscripts/fm11rf08_full.py
Comment (doegox) [inline client/pyscripts/fm11rf08_full.py:1]: I find the script name confusing.
fm11rf08s_recovery provides also a full dump and a key dump, like autopwn
so call this one "demo" or "verbose" but not  "full"

File: client/pyscripts/fm11rf08_full.py
Comment (doegox) [inline client/pyscripts/fm11rf08_full.py:221]: it would be interesting to isolate this function in a separate file to be imported, so other ppl can use it in their scripts s well, or call the script with a provided block0

File: client/pyscripts/fm11rf08_full.py
Comment (doegox) [inline client/pyscripts/fm11rf08_full.py:296]: it would be interesting to isolate this function in a separate file to be imported, so other ppl can use it in their scripts s well, or call the script with a provided block0

File: client/pyscripts/fm11rf08_full.py
Comment (doegox) [inline client/pyscripts/fm11rf08_full.py:464]: redundant with the dump file already provided by recovery()
or call it verifyDump as you have the verifyKeys...

File: client/pyscripts/fm11rf08_full.py
Comment (doegox) [inline client/pyscripts/fm11rf08_full.py:511]: note that dump from recovery is already patched

File: client/pyscripts/fm11rf08_full.py
Comment (doegox) [inline client/pyscripts/fm11rf08_full.py:570]: it would be interesting to isolate this function in a separate file to be imported, so other ppl can use it in their scripts s well, or call the script with a provided block0

File: client/pyscripts/fm11rf08_full.py
Comment (doegox) [inline client/pyscripts/fm11rf08_full.py:602]: it would be interesting to isolate this function in a separate file to be imported, so other ppl can use it in their scripts s well, or call the script with a provided block0

File: client/pyscripts/fm11rf08_full.py
Comment (doegox) [inline client/pyscripts/fm11rf08_full.py:793]: it would be interesting to isolate this function in a separate file to be imported, so other ppl can use it in their scripts s well, or call the script with a provided block0.
Note that we already have `hf mf acl`

File: client/pyscripts/fm11rf08_full.py
Comment (csBlueChip) [reply client/pyscripts/fm11rf08_full.py:1]: if you dump the card using `recovery` and then dump the card using `full` and then do an `ls` you will see that the (so called) "full" file is 128 bytes longer because (as discussed in the verbose intro) it includes "all 8 documented Dark blocks"
```
proxmark@linux:~$ ls hf-mf-5CB49CA6-dump*.bin -l
-rw-r--r-- 1 prox prox    1152    Nov  4 00:26 hf-mf-5CB49CA6-dump18.bin
-rw-r--r-- 1 prox prox    1024    Oct 30 18:59 hf-mf-5CB49CA6-dump.bin
```
As for the name ...It was called `bc.py` until the last minute, and I just thought "the full 18 blocks" is what I was going after, and grabbed the word "full" ...If you think that name is misleading, then yes, let's pick something better.

File: client/pyscripts/fm11rf08_full.py
Comment (csBlueChip) [reply client/pyscripts/fm11rf08_full.py:87]: Again I just wanted a way to highlight that this script retrieves the 8 Dark blocks, I am happy to accept a better/clearer message that conveys that.

File: client/pyscripts/fm11rf08_full.py
Comment (csBlueChip) [reply client/pyscripts/fm11rf08_full.py:221]: If/when this script gets pulled, I would be flattered if functions were yanked to a library. If/when that happens it would probably be wise to pick a more "API" style name for the function. I dunno, maybe `hf_mf_block0_decode(binaryData[16])` ...but I'm sure you guys probably have a naming convention - so "whatever fits" is good by me

Same goes for the following 3 similar comments :)

File: client/pyscripts/fm11rf08_full.py
Comment (csBlueChip) [reply client/pyscripts/fm11rf08_full.py:296]: qv. https://github.com/RfidResearchGroup/proxmark3/pull/2613#discussion_r1830270895

File: client/pyscripts/fm11rf08_full.py
Comment (csBlueChip) [reply client/pyscripts/fm11rf08_full.py:570]: qv. https://github.com/RfidResearchGroup/proxmark3/pull/2613#discussion_r1830270895

File: client/pyscripts/fm11rf08_full.py
Comment (csBlueChip) [reply client/pyscripts/fm11rf08_full.py:602]: qv. https://github.com/RfidResearchGroup/proxmark3/pull/2613#discussion_r1830270895

File: client/pyscripts/fm11rf08_full.py
Comment (csBlueChip) [reply client/pyscripts/fm11rf08_full.py:464]: Considerations:
~ This function reads the 8 Dark blocks in addition to the 64 "normal" blocks (which were read by `recovery()`)
~ `verifyKeys` takes the known keys and verifies that they are still valid. ...because we would like to avoid waiting "<30mins" if possible
~ `readBlocks` deliberately re-reads the card each time it is run, so as to detect edits to the card ...it only takes a few seconds ...I have to read it all to verify nothing has changed, at which point, well, I've just read it all ...There may well be credit to writing a `verifyBlocks()` function to detect changes - but that would need to call `readblock()` and `loadOldFile()`
~ The feature bloat had already crept in when I started detecting and decoding various tags - I just had to draw the line somewhere (for now)

File: client/pyscripts/fm11rf08_full.py
Comment (csBlueChip) [reply client/pyscripts/fm11rf08_full.py:793]: qv. https://github.com/RfidResearchGroup/proxmark3/pull/2613#discussion_r1830270895

this function requires data other than block 0

I was originally going to call `hf mf acl` (as I say, this script is initially to support a defcon workshop) and parse the output (if I had to), but in the end I found the output difficult to decipher, and tricky to parse - in the end, some bit twiddling just seemed easier.

File: client/pyscripts/fm11rf08_full.py
Comment (csBlueChip) [reply client/pyscripts/fm11rf08_full.py:511]: For reasons discussed above, I do not load the existing `dump.bin` so this action is 'required'

**Comment by doegox** [comment]
> Thank you for taking time to explain your position. I really hadn't considered the potential nightmare of maintaining multiple licences in one repository. I totally understand.

Just to say, I second Iceman comment on the license with this example: I discussed many parts of this script as better being moved into some common MFC helper functions Python module. Now if each helper function had to retain its original license, it becomes a nightmare to constitute such module mixing licenses.

> In case you're wondering: I have both personal and professional reasons to dislike what GPL3 has done to hinder development of/with cross-platform support, and the damage it has done to the security landscape; and I just want to make sure anything I personally release can be used by anybody for any reason and in any way they so desire ...yourself (and this fantastic project) included :)

Fair enough. On our side, we tend to dislike the several manufacturers who stole the proxmark3 code to make their closed-source variants...
 
> How about this?: ~ remove the `print("Licence....")` line from the code ~ add this comment to the top of the script "To minimise administrative overheads on the Proxmark code, this copy of the script is released under the GPLv3 licence. If you require/desire a less restrictive licence, this script may instead be obtained from [a link on my github page] under the MIT licence" ...Does that work for you? :)

IMHO the first sentence is useless.
"The original version of this script is also kept at your disposal under the MIT license at the following link..."

"Original version" is important because once merged, over time, this code will evolve in the repo.
E.g. we can merge it now and tomorrow I'll do the code refactoring into Python module(s).

File: client/pyscripts/fm11rf08_full.py
Comment (doegox) [reply client/pyscripts/fm11rf08_full.py:1]: Ha! Initially I didn't make the link when you say "dark".
You use the term "dark" both for the backdoor key (see getDarkKey) and for the advanced verification method blocks.
In MFC, dark refers to darkside (and now the variant darknested which replaces darkside+nested)
and personally I would avoid reusing the same term in other contexts.
* dark key => backdoor key
* dark blocks => advanced verification method blocks (or some shorter avblocks for example)
If the extra blocks are useful to get, better to change the existing commands than adding an entire new script.
Right now I refrained to do so because the existing commands cannot handle dumps of unexpected sizes, so your dump cannot be used by e.g. hf mf restore. We already have the problem with the keys file at the moment.
This is planned to handle properly these oversized dumps and keyfiles, but it takes time.
So we can merge your code but don't be surprised if tomorrow it becomes redundant.

File: client/pyscripts/fm11rf08_full.py
Comment (doegox) [reply client/pyscripts/fm11rf08_full.py:464]: if you need to read the card, using "hf mf ecfill -c 4 -k ..." is wayyy faster.
Then, yes, for now, you've to dump manually the 8 avblocks because as said above, it is not yet supported in ecfill or any other mfc emulation command. (I think even letting a "hf mf ecfill --4k -c 4 -k..." dumping and failing over unexisting sectors is still faster)

**Review by doegox — APPROVED** [review]

**Comment by iceman1001** [comment]
Here is what I suggest you put on top of your script file.   
And remove the line in Main().

> #
> #  Copyright @csBlueChip
> #
> #   This program is free software: you can redistribute it and/or modify
> #   it under the terms of the GNU General Public License as published by
> #   the Free Software Foundation, either version 3 of the License, or
> #   (at your option) any later version.
> #
> #   This program is distributed in the hope that it will be useful,
> #   but WITHOUT ANY WARRANTY; without even the implied warranty of
> #   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
> #   GNU General Public License for more details.
> #
> #   See LICENSE.txt for the text of the license.
> #
> #  The original version of this script is also kept at your disposal under the MIT license at the following link   <enter link>
> #   or reach out to author for alternative licenses. 
> 

and since we use git,  we don't need a "internal" change log.


When @doegox   is happy,   we merge

**Comment by doegox** [comment]
We can merge, then I'll make the discussed changes (module etc)

**Comment by doegox** [comment]
but yeah change the licensing terms before merge, else it looks like maintainers are rewriting your terms...

File: client/pyscripts/fm11rf08_full.py
Comment (csBlueChip) [reply client/pyscripts/fm11rf08_full.py:1]: As I say, the script as it stands solved a very specific problem. And, as with the "library functions" comments, if this script become a part of the (metaphoric) Borg Collective, that's no bad thing.

Re. the word "dark": We needed a word so we could discuss it. Your paper used the phrase "dark keys" and extended nested to "dark nested" ...So we took inspiration, and called the additional blocks (which we discovered from the same paper) "dark blocks" ...Again, if it's confusing - this is exactly the right moment to change it :)

**Comment by csBlueChip** [comment]
> but yeah change the licensing terms before merge, else it looks like maintainers are rewriting your terms...

it is done

**Comment by iceman1001** [comment]
Excellent!
I like the idea for your ctf challenge

**Pull request merged**
